### PR TITLE
Source test

### PR DESCRIPTION
**Pull request by gurghet**
This adds some tests for the source method which is at the core of Tamer.
Please note that the tests are biased towards characterizing what is already there and the bugs that we have to deal with, there is not a lot in term of complex edge cases.
Also the automatic tape rewinding is not tested, only the manual strategy is.
This is done to lay the foundation of further work on testing and in case we want to implement workarounds.

### CODE REVIEW

**Review by barambani — APPROVED** [review]

File: core/src/test/scala/tamer/SinkSpec.scala [hunk 1]
```diff
@@ -2,11 +2,10 @@ package tamer
 
 import io.confluent.kafka.schemaregistry.ParsedSchema
 import log.effect.zio.ZioLogWriter.log4sFromName
-import org.apache.kafka.clients.producer.{ProducerRecord, RecordMetadata}
-import org.apache.kafka.common.{Metric, MetricName, TopicPartition}
+import org.apache.kafka.clients.producer.ProducerRecord
+import utils.{FailingFakeProducer, FakeProducer}
```
Comment (sirocchj) [inline core/src/test/scala/tamer/SinkSpec.scala:6]: if we namespace these we want to use the fully qualified package name here, ie
```suggestion
import tamer.utils.{FailingFakeProducer, FakeProducer}
```
alternatively, we can leave both of the above in the main package

File: core/src/test/scala/tamer/TamerSpec.scala [hunk 2]
```diff
@@ -52,7 +52,7 @@ object TamerSpec extends DefaultRunnableSpec with TamerSpecGen {
 
         val p1Offset = fixedPartitions(AdminPartition(p))
 
-        assert(p1Offset.offset)(equalTo(12L))
+        assert(p1Offset.offset)(equalTo(12L)) &&
```
Comment (sirocchj) [inline core/src/test/scala/tamer/TamerSpec.scala:55]: 💯

File: core/src/test/scala/utils/FakeConsumer.scala [hunk 3]
```diff
@@ -0,0 +1,164 @@
+package utils
+
+import log.effect.LogWriter
+import org.apache.kafka.clients.consumer.{ConsumerRecord, OffsetAndMetadata, OffsetAndTimestamp}
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.common.{Metric, MetricName, PartitionInfo, TopicPartition}
+import utils.FakeConsumerUtils.{getCommittedOr0, increaseForPartition}
```
Comment (sirocchj) [inline core/src/test/scala/utils/FakeConsumer.scala:7]: same story re fully qualified package

File: build.sbt [hunk 4]
```diff
@@ -108,6 +109,7 @@ lazy val core = project
       "io.laserdisc"                     %% "log-effect-zio"                 % V.`log-effect`,
       "is.cir"                           %% "ciris"                          % V.ciris,
       "org.apache.kafka"                  % "kafka-clients"                  % V.kafka,
+      "org.scalatestplus"                %% "scalatestplus-mockito"          % V.mockito       % Test,
```
Comment (sirocchj) [inline build.sbt:112]: can we get rid of this at next iteration? I believe it's only needed to mock out the recent addition of the AdminClient in the core

**Review by sirocchj — APPROVED** [review]
Minor comments, can be addressed separately too

File: core/src/test/scala/tamer/SinkSpec.scala (hunk 1 above)
Comment (gurghet) [reply core/src/test/scala/tamer/SinkSpec.scala:6]: utils => tamer.utils

File: build.sbt (hunk 4 above)
Comment (gurghet) [reply build.sbt:112]: Yeah, I surely hate it but there are several direction we could go towards. Do we want also a fake admin so we can test the automatic rewind? I tried and it wasn't trivial so I dropped it, but it's certainly possible. Do we want to just remove the dependency? In this case we need a homebrew mock, which won't be easy given that, unlike producer and consumer, this doesn't have an interface.

Comment (sirocchj) [reply build.sbt:112]: I'd go with the former, but aside for not being trivial I guess we also need to abstract away the construction of the `AdminClient` ourselves in `tamer.Tamer`, so we can replace this https://github.com/zio/zio-kafka/blob/007bebb58d8542e4c49bd6078b3e9bdea559d12e/src/main/scala/zio/kafka/admin/AdminClient.scala#L625-L630 with our FakeAdminClient, right?

Let's think about it a bit more and maybe create a card for it?

**Review by sirocchj — APPROVED** [review]

File: build.sbt (hunk 4 above)
Comment (gurghet) [reply build.sbt:112]: Ok let's also talk about what kind of abstraction we want. I would also evaluate directly contributing by enveloping the AdminClient into service so that everything become uniform and homogeneous.

Comment (sirocchj) [reply build.sbt:112]: Agreed. We like for sure the idea of having an `Admin` service with a live implementation backed by the current java AdminClient, so let's try and contribute to that!
